pkg/controllers/management/aks/aks_cluster_handler.go

```diff
@@ -0,0 +1,696 @@
+		if err != nil {
+			return cluster, err
+		}
+
+	}
+
+	aksClusterConfigMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&cluster.Spec.AKSConfig)
+	if err != nil {
+		return cluster, err
+	}
+
+	// check for changes between aks spec on cluster and the aks spec on the aksClusterConfig object
+	if !reflect.DeepEqual(aksClusterConfigMap, aksClusterConfigDynamic.Object["spec"]) {
+		logrus.Infof("change detected for cluster [%s], updating AKSClusterConfig", cluster.Name)
+		return e.updateAKSClusterConfig(cluster, aksClusterConfigDynamic, aksClusterConfigMap)
+	}
+
+	// get aks Cluster Config's phase
+	status, _ := aksClusterConfigDynamic.Object["status"].(map[string]interface{})
+	phase, _ := status["phase"]
+	failureMessage, _ := status["failureMessage"].(string)
+
+	switch phase {
+	case "creating":
+		// set provisioning to unknown
+		cluster, err = e.setUnknown(cluster, apimgmtv3.ClusterConditionProvisioned, "")
+		if err != nil {
+			return cluster, err
+		}
+
+		if cluster.Status.AKSStatus.UpstreamSpec == nil {
+			cluster, err = e.setInitialUpstreamSpec(cluster)
+			if err != nil {
+				return cluster, err
+			}
+			return cluster, nil
+		}
+
+		e.clusterEnqueueAfter(cluster.Name, enqueueTime)
+		if failureMessage == "" {
+			logrus.Infof("waiting for cluster AKS [%s] to finish creating", cluster.Name)
+			return e.setUnknown(cluster, apimgmtv3.ClusterConditionProvisioned, "")
+		}
+		logrus.Infof("waiting for cluster AKS [%s] create failure to be resolved", cluster.Name)
+		return e.setFalse(cluster, apimgmtv3.ClusterConditionProvisioned, failureMessage)
+	case "active":
+		if cluster.Status.AKSStatus.UpstreamSpec == nil {
+			cluster, err = e.setInitialUpstreamSpec(cluster)
+			if err != nil {
+				return cluster, err
+			}
+			return cluster, nil
+		}
+
+		if apimgmtv3.ClusterConditionPending.IsUnknown(cluster) {
+			cluster = cluster.DeepCopy()
+			apimgmtv3.ClusterConditionPending.True(cluster)
+			cluster, err = e.clusterClient.Update(cluster)
+			if err != nil {
+				return cluster, err
+			}
+		}
+
+		if cluster.Spec.AKSConfig.Imported {
+			if cluster.Status.AKSStatus.UpstreamSpec == nil {
+				// non imported clusters will have already had upstream spec set
+				return e.setInitialUpstreamSpec(cluster)
+			}
+
+			if apimgmtv3.ClusterConditionPending.IsUnknown(cluster) {
+				cluster = cluster.DeepCopy()
+				apimgmtv3.ClusterConditionPending.True(cluster)
+				cluster, err = e.clusterClient.Update(cluster)
+				if err != nil {
+					return cluster, err
+				}
+			}
+		}
+
+		addNgMessage := "Cannot deploy agent without NodePools. Add a NodePool."
+		noNodePoolsOnSpec := len(cluster.Spec.AKSConfig.NodePools) == 0
+		noNodePoolsOnUpstreamSpec := len(cluster.Status.AKSStatus.UpstreamSpec.NodePools) == 0
+		if (cluster.Spec.AKSConfig.NodePools != nil && noNodePoolsOnSpec) || (cluster.Spec.AKSConfig.NodePools == nil && noNodePoolsOnUpstreamSpec) {
```

Comment:
Is this needed? Isn't it already one of the validators? I see this in the EKS handler, does it apply the same way to AKS?

Reply:
fixed